stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,787 @@
+                          relative_tolerance_forward_);
+    check_positive_finite(function_name, "absolute_tolerance_forward",
+                          absolute_tolerance_forward_);
+    check_size_match(function_name, "absolute_tolerance_forward",
+                     absolute_tolerance_forward_.size(), "states", N_);
+    check_positive_finite(function_name, "relative_tolerance_backward",
+                          relative_tolerance_backward_);
+    check_positive_finite(function_name, "absolute_tolerance_backward",
+                          absolute_tolerance_backward_);
+    check_size_match(function_name, "absolute_tolerance_backward",
+                     absolute_tolerance_backward_.size(), "states", N_);
+    check_positive_finite(function_name, "relative_tolerance_quadrature",
+                          relative_tolerance_quadrature_);
+    check_positive_finite(function_name, "absolute_tolerance_quadrature",
+                          absolute_tolerance_quadrature_);
+    check_positive(function_name, "max_num_steps", max_num_steps_);
+    check_positive(function_name, "num_steps_between_checkpoints",
+                   num_steps_between_checkpoints_);
+    // for polynomial: 1=CV_HERMITE / 2=CV_POLYNOMIAL
+    check_bounded(function_name, "interpolation_polynomial",
+                  interpolation_polynomial_, 1, 2);
+    // 1=Adams=CV_ADAMS, 2=BDF=CV_BDF
+    check_bounded(function_name, "solver_forward", solver_forward_, 1, 2);
+    check_bounded(function_name, "solver_backward", solver_backward_, 1, 2);
+
+    solver_ = new cvodes_solver(
+        function_name, f, N_, num_args_vars_, ts_.size(), solver_forward_,
+        state_forward_, state_backward_, quad_, absolute_tolerance_forward_,
+        absolute_tolerance_backward_);
+
+    check_flag_sundials(
+        CVodeInit(solver_->cvodes_mem_, &cvodes_integrator_adjoint_vari::cv_rhs,
+                  value_of(t0_), solver_->nv_state_forward_),
+        "CVodeInit");
+
+    // Assign pointer to this as user data
+    check_flag_sundials(
+        CVodeSetUserData(solver_->cvodes_mem_, reinterpret_cast<void*>(this)),
+        "CVodeSetUserData");
+
+    cvodes_set_options(solver_->cvodes_mem_, relative_tolerance_forward_,
+                       absolute_tolerance_forward_(0), max_num_steps_);
+
+    check_flag_sundials(
+        CVodeSVtolerances(solver_->cvodes_mem_, relative_tolerance_forward_,
+                          solver_->nv_absolute_tolerance_forward_),
+        "CVodeSVtolerances");
+
+    check_flag_sundials(
+        CVodeSetLinearSolver(solver_->cvodes_mem_, solver_->LS_forward_,
+                             solver_->A_forward_),
+        "CVodeSetLinearSolver");
+
+    check_flag_sundials(
+        CVodeSetJacFn(solver_->cvodes_mem_,
+                      &cvodes_integrator_adjoint_vari::cv_jacobian_rhs_states),
+        "CVodeSetJacFn");
+
+    // initialize backward sensitivity system of CVODES as needed
+    if (is_var_return_) {
```

Comment:
I think we also don't need the adjoint solve if only ts are vars, but if it makes the memory management easier totally fine to leave this as is.

Reply:
no, it's correct that if only ts is a var, we do not need the backward system ever be initialized. I just found that the if logic is more than I could compile in my head quickly enough such that I thought.. let's just keep it simple. Having only ts as a var is anyway an odd edge case which I don't expect to matter (one should anyway use the other solvers presumably).